cmd/snap/cmd_interface.go

```diff
@@ -0,0 +1,183 @@
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"sort"
+	"text/tabwriter"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+
+	"github.com/jessevdk/go-flags"
+)
+
+type cmdInterface struct {
+	ShowAttrs   bool `long:"attrs"`
+	ShowAll     bool `long:"all"`
+	Positionals struct {
+		Interface interfaceName `skip-help:"true"`
+	} `positional-args:"true"`
+}
+
+var shortInterfaceHelp = i18n.G("Lists snap interfaces")
+var longInterfaceHelp = i18n.G(`
+The interface command shows details of snap interfaces.
+
+If no interface name is provided, a list of interface names with at least
+one connection is shown, or a list of all interfaces if --all is provided.
+`)
+
+func init() {
+	addCommand("interface", shortInterfaceHelp, longInterfaceHelp, func() flags.Commander {
+		return &cmdInterface{}
+	}, map[string]string{
+		"attrs": i18n.G("Show interface attributes"),
+		"all":   i18n.G("Include unused interfaces"),
+	}, []argDesc{{
+		name: i18n.G("<interface>"),
+		desc: i18n.G("Show details of a specific interface"),
+	}})
+}
+
+func (x *cmdInterface) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	if x.Positionals.Interface != "" {
+		// Show one interface in detail.
+		name := string(x.Positionals.Interface)
+		ifaces, err := Client().Interfaces(&client.InterfaceQueryOptions{
+			Names: []string{name},
+			Doc:   true,
+			Plugs: true,
+			Slots: true,
+		})
+		if err != nil {
+			return err
+		}
+		if len(ifaces) == 0 {
+			return fmt.Errorf(i18n.G("no such interface"))
+		}
+		x.showOneInterface(ifaces[0])
+	} else {
+		// Show an overview of available interfaces.
+		ifaces, err := Client().Interfaces(&client.InterfaceQueryOptions{
+			Connected: !x.ShowAll,
+		})
+		if err != nil {
+			return err
+		}
+		if len(ifaces) == 0 {
+			if x.ShowAll {
+				return fmt.Errorf(i18n.G("no interfaces found"))
+			}
+			return fmt.Errorf(i18n.G("no interfaces currently connected"))
+		}
+		x.showManyInterfaces(ifaces)
+	}
+	return nil
+}
+
+func (x *cmdInterface) showOneInterface(iface *client.Interface) {
+	w := tabwriter.NewWriter(Stdout, 2, 2, 1, ' ', 0)
+	defer w.Flush()
+
+	fmt.Fprintf(w, "name:\t%s\n", iface.Name)
+	if iface.Summary != "" {
+		fmt.Fprintf(w, "summary:\t%s\n", iface.Summary)
+	}
+	if iface.DocURL != "" {
+		fmt.Fprintf(w, "documentation:\t%s\n", iface.DocURL)
+	}
+	if len(iface.Plugs) > 0 {
+		fmt.Fprintf(w, "plugs:\n")
+		for _, plug := range iface.Plugs {
+			if plug.Name == iface.Name {
+				fmt.Fprintf(w, "  - %s", plug.Snap)
+			} else {
+				fmt.Fprintf(w, `  - "%s:%s"`, plug.Snap, plug.Name)
+			}
+			// Print a colon which will make the snap:plug element a key-value
+			// yaml object so that we can write the label or attributes.
+			if plug.Label != "" || len(plug.Attrs) > 0 && x.ShowAttrs {
+				fmt.Fprintf(w, ":\n")
+			} else {
+				fmt.Fprintf(w, "\n")
+			}
+			if plug.Label != "" {
+				fmt.Fprintf(w, "      label:\t%s\n", plug.Label)
```

Comment:
Oh, even better. Thanks!

I think we can still stick key-value attributes below (for when we actually have any).